source/simulator/core.cc

```diff
@@ -2147,6 +2162,329 @@ namespace aspect
+
+                  /**
+                   * This method allows to slowly introduce the derivatives based
+                   * on the improvement of the residual. If we do not use it, we
+                   * just set it so the newton_derivative_scaling_factor goes from
+                   * zero to one when switching on the Newton solver.
+                   */
+                  if (!parameters.use_newton_residual_scaling_method)
+                    parameters.newton_residual = 0;
+                }
+
+              parameters.newton_derivative_scaling_factor = std::max(0.0,(1.0-(parameters.newton_residual/parameters.switch_initial_newton_residual)));
+
+
+              /**
+               * copied from solver.cc
+               */
+
+              // Many parts of the solver depend on the block layout (velocity = 0,
+              // pressure = 1). For example the linearized_stokes_initial_guess vector or the StokesBlock matrix
+              // wrapper. Let us make sure that this holds (and shorten their names):
+              const unsigned int block_vel = introspection.block_indices.velocities;
+              const unsigned int block_p = (parameters.include_melt_transport) ?
+                                           introspection.variable("fluid pressure").block_index
+                                           : introspection.block_indices.pressure;
+              Assert(block_vel == 0, ExcNotImplemented());
+              Assert(block_p == 1, ExcNotImplemented());
+              Assert(!parameters.include_melt_transport
+                     || introspection.variable("compaction pressure").block_index == 1, ExcNotImplemented());
+
+              // create a completely distributed vector that will be used for
+              // the scaled and denormalized solution and later used as a
+              // starting guess for the linear solver
+              LinearAlgebra::BlockVector linearized_stokes_initial_guess (introspection.index_sets.stokes_partitioning, mpi_communicator);
+
+              linearized_stokes_initial_guess.block (block_vel) = current_linearization_point.block (block_vel);
+              linearized_stokes_initial_guess.block (block_p) = current_linearization_point.block (block_p);
+
+              if (nonlinear_iteration == 0)
+                {
+                  current_linearization_point.block(introspection.block_indices.velocities) = 0;
+                  linearized_stokes_initial_guess.block (block_vel) = 0;
+
+                  rebuild_stokes_matrix = assemble_newton_stokes_system = true;
+                  rebuild_stokes_preconditioner = false;
+
+
+                  denormalize_pressure (this->last_pressure_normalization_adjustment,
+                                        linearized_stokes_initial_guess,
+                                        current_linearization_point);
+
+                  compute_current_constraints ();
+                  assemblers.reset (new internal::Assembly::AssemblerLists<dim>());
```

Comment:
why is this reset and the call to `set_assembers` in the next line necessary? These function should be called once at the beginning of the model to select the appropriate assemblers. Or is this necessary because of the switch from Picard to Newton? Then this call should only happen when we just switched (i.e. around line 2244)